### ISSUE TITLE
Firebase remote config fetch failure causes `WifiDetailsRepoImpl` not to return default or cached values for wifi details

### ISSUE THREAD

**Issue by johnGachihi** [droidconKE/droidconKE2019App#72]
I noticed that when a firebase remote config (frc) fetch fails in [`WifiDetailsRepoImpl`](https://github.com/droidconKE/droidconKE2019App/blob/470bc1669444c7d66a89986e27230c23093e7a14/app/src/main/java/com/android254/droidconke19/repository/WifiDetailsRepo.kt#L19), instead of falling back to and returning the default or cached frc wifi details, FirebaseResult.error() is returned.
This happens because the `await()` ext. function throws an exception whenever the fetch fails. And since it is wrapped in [`runCatching`](https://github.com/droidconKE/droidconKE2019App/blob/develop/app/src/main/java/com/android254/droidconke19/datastates/RunCatching.kt), a `FirebaseResult.error()` is returned.

Ideally, whenever an frc fetch fails, the frc defaults or cached values should be returned.

My approach to solve this would be to wrap [this](https://github.com/droidconKE/droidconKE2019App/blob/470bc1669444c7d66a89986e27230c23093e7a14/app/src/main/java/com/android254/droidconke19/repository/WifiDetailsRepo.kt#L19) in a try-catch like so:
```
try {
      firebaseRemoteConfig.fetchAndActivate().await()
} catch (e: Exception) {}

```
so that even when the fetch fails this will be called

`WifiDetailsModelFactory.create(firebaseRemoteConfig)`

and return the default or cached values

**Comment by wangerekaharun** [comment]
Will check on this and see the best possible way to handle it with your suggestion in mind

**Comment by johnGachihi** [comment]
We could also use kotlin Flow to first emit the cached data then (in the background) refresh the cached data by fetching and emit the data again if it changes. This way even if the fetch fails the UI will already contain the cached data
```
// In WifiDetailsRepo
override fun fetchWifiDetails(): Flow<FirebaseResult<WifiDetailsModel>> = flow {
    val cachedTravelInfo = runCatching { TravelInfoModelFactory.create(firebaseRemoteConfig) }
    emit(cachedTravelInfo)

    val refreshedTravelInfo = runCatching {
      firebaseRemoteConfig.fetchAndActivate().await()
      TravelInfoModelFactory.create(firebaseRemoteConfig)
    }

    if (cachedTravelInfo != refreshedTravelInfo)
       emit(freshTravelInfo)
}


// In the ViewModel
fun getWifiDetails() = viewModelScope.launch(Dispatchers.IO) {
    wifiDetailsRepo.fetchWifiDetails().collect { value ->
      when (value) {
        is FirebaseResult.Success -> wifiDetails.postValue(value.data)
        is FirebaseResult.Error -> fetchError.postValue(value.exception)
      }
    }
}
```

**Comment by wangerekaharun** [comment]
@johnGachihi  Anyway to reproduce the said error?

**Comment by johnGachihi** [comment]
Disconnect your phone from internet
Then open the info page

**Comment by wangerekaharun** [comment]
Okay will test this and get back to you

**Comment by wangerekaharun** [comment]
On this case, am trying to investigate why it's not loading the default values in case of an error at the very least.

**Comment by michaelbukachi** [comment]
RemoteConfig doesn't work sometimes. I have faced this issue before.

**Comment by wangerekaharun** [comment]
Okay we can go with what @johnGachihi  is suggesting

**Comment by michaelbukachi** [comment]
Haha. Flow to the rescue once again.

**Comment by wangerekaharun** [comment]
> Haha. Flow to the rescue once again.
@johnGachihi you can do that pull request. 

Here are my thoughts on the implementation on [EventTypeViewModel](https://github.com/droidconKE/droidconKE2019App/blob/develop/app/src/main/java/com/android254/droidconke19/viewmodels/EventTypeViewModel.kt). I think the `also` function is not necessary on the `wifiDetails` variable.

```

    val wifiDetails: MutableLiveData<FirebaseResult<WifiDetailsModel>> by lazy {
        MutableLiveData<FirebaseResult<WifiDetailsModel>>().also {
            fetchWifiDetails()
        }
    }
```
Or what was your thinking in making it to be this way?

**Comment by johnGachihi** [comment]
Without it, the method `fetchWifiDetails()` would have to be called in `EventFragment` (as with `fetchSession()`) .
So I saw that it will be called every time `EventFragment` is recreated.

**Comment by wangerekaharun** [comment]
> Without it, the method `fetchWifiDetails()` would have to be called in `EventFragment` (as with `fetchSession()`) .
> So I saw that it will be called every time `EventFragment` is recreated.

Yes am aware of that #73 is aimed at improving that and avoid the methods being called on fragment recreation. Thank you for your clarification!

**Comment by johnGachihi** [comment]
> On this case, am trying to investigate why it's not loading the default values in case of an error at the very least.

`firebaseRemoteConfig.fetchAndActivate().await()`
When the `fetchAndActivate()` fails, `await()` throws an exception and since the line above is called inside `runCatching {  }`, `FirebaseResult.Error(`exception thrown`)` is returned.

**Comment by johnGachihi** [comment]
Sorry, closed it by mistake

**Comment by wangerekaharun** [comment]
I have re-done the TravelFragment Remote config architecture, with what we were following previously with the app and it works without showing the error. Seems the problem is with your implementation bit. Let me do re-write and see if the error persists

**Comment by wangerekaharun** [comment]
@johnGachihi Latest commit fixes the issue

**Comment by johnGachihi** [comment]
@wangerekaharun, in `TravelDetailsRepo`, you left out `await()` after `fetchAndActivate()`.
It's the one that was causing the issue in my code
